Add error note for failed coercions to optional types and error unions

WillLillis opened the pull request:
This adds a helpful error note in cases where coercion to an optional type fails.

Using the example given from #20186, here's a before:

```shell
x.zig:2:20: error: expected type '?u8', found 'comptime_int'
    const x: ?u8 = 256;
                   ^~~
```

And after:

```shell
src/x.zig:20:20: error: expected type '?u8', found 'comptime_int'
    const x: ?u8 = 256;
                   ^~~
src/x.zig:20:20: note: child type 'u8' cannot represent value '256'
```

Another option for this case would be to replace the original error message instead of adding a note, as suggested in #20186 maybe with something along the lines of 

```shell
src/x.zig:20:20: error: type '?u8' cannot represent value '256'
```

I'm still relatively new to the project, so if I've missed something my apologies for the noise.

Closes #20186

nektro commented:
excellent work! will need a test though; best place for this kind of one is in `/test/cases/compile_errors`

running `stage4/bin/zig build test-cases -Dskip-non-native -Dskip-release` will help you know it worked

WillLillis commented:
> excellent work! will need a test though; best place for this kind of one is in `/test/cases/compile_errors`
> 
> running `stage4/bin/zig build test-cases -Dskip-non-native -Dskip-release` will help you know it worked

Awesome, thanks so much for the help! 

I had to update the expected output of two other tests with this change, I'm assuming that's ok?

nektro commented:
yep!

Vexu commented on test/cases/compile_errors/any_typed_null_to_any_typed_optional.zig:
```diff
@@ -8,4 +8,5 @@ pub export fn entry() void {
 // target=native
 //
 // :3:9: error: expected type '?*anyopaque', found '?usize'
+// :3:9: note: child type '*anyopaque' cannot represent value 'null'
```
The value being `null` has nothing to do with why this coercion failed.

Vexu commented on test/cases/compile_errors/type_mismatch_in_C_prototype_with_varargs.zig:
```diff
@@ -11,5 +11,6 @@ export fn main() void {
 // target=native
 //
 // :5:22: error: expected type '?fn ([*c]u8, ...) callconv(.C) void', found 'fn ([*:0]u8, ...) callconv(.C) void'
+// :5:22: note: child type 'fn ([*c]u8, ...) callconv(.C) void' cannot represent value '(extern function 'fn_decl')'
```
Same thing with this note.

Vexu reviewed (changes requested):
I think this would be better handled like `double_ptr_to_anyopaque` was by adding more cases to `InMemoryCoercionResult` (which should probably have been renamed to just `CoercionResult`).

I think it should be safe to add a `dest_tag == .Int and src_tag == .ComptimeInt` case to `coerceInMemoryAllowed`. This would have the additional benefit of also working for error unions.

WillLillis commented:
Thanks so much for taking a look! 

> I think this would be better handled like `double_ptr_to_anyopaque` was by adding more cases to `InMemoryCoercionResult` (which should probably have been renamed to just `CoercionResult`).
> 
> I think it should be safe to add a `dest_tag == .Int and src_tag == .ComptimeInt` case to `coerceInMemoryAllowed`. This would have the additional benefit of also working for error unions.

I refactored the check to work as another case in `InMemoryCoercionResult` , and added the  `dest_tag == .Int and src_tag == .ComptimeInt` case to `coerceInMemoryAllowed`. I wasn't able to work out a good way to get the actual problematic value to print in the error message after these changes. I'm probably missing something obvious. I wasn't sure what you meant for the error union case, so I held off on trying to write a test for that yet.

> `InMemoryCoercionResult` (which should probably have been renamed to just `CoercionResult`).

Would it be worth changing the name in this PR or in a followup?

Vexu commented:
> I wasn't able to work out a good way to get the actual problematic value to print in the error message after these changes. I'm probably missing something obvious. 

It's definitely not obvious. One possible solution I could think of is adding a `error_details: *?InMemoryCoercionResult` field to `CoerceOpts` and populating that at the `type '{}' cannot represent integer value '{}'` error.

Simpler alternative would probably be to pass an optional `src_val: ?Value` to `coerceInMemoryAllowed` and checking that instead of calling `coerceExtra`.

> I wasn't sure what you meant for the error union case, so I held off on trying to write a test for that yet.

Something like:
```zig
export fn b() void {
    const x: anyerror!u8 = 256;
    _ = x;
}
```

> Would it be worth changing the name in this PR or in a followup?

Different PR definitely.

---
And sorry for taking so long to answer.

WillLillis commented:
> > I wasn't able to work out a good way to get the actual problematic value to print in the error message after these changes. I'm probably missing something obvious.
> 
> It's definitely not obvious. One possible solution I could think of is adding a `error_details: *?InMemoryCoercionResult` field to `CoerceOpts` and populating that at the `type '{}' cannot represent integer value '{}'` error.

I spent some time trying to get this approach working but wasn't able to. It seems like the call I added from `coerceInMemoryAllowed` to `coerceExtra` doesn't directly reach the comptime int to int conversion check, so any pointer I passed through `opts` didn't make it to the point where I wanted to populate it. 

> Simpler alternative would probably be to pass an optional `src_val: ?Value` to `coerceInMemoryAllowed` and checking that instead of calling `coerceExtra`.

I tried this out and it worked!

> > I wasn't sure what you meant for the error union case, so I held off on trying to write a test for that yet.
> 
> Something like:
> 
> ```zig
> export fn b() void {
>     const x: anyerror!u8 = 256;
>     _ = x;
> }
> ```

Gotcha. It seems like that case isn't caught with my current changes, but I'll be able to spend some more time on this tomorrow/ later this week as well.

> And sorry for taking so long to answer.

No worries at all! I know this is low priority relative to everything else going on in the project. Really appreciate the feedback regardless of timing.

Vexu commented on src/Sema.zig:
```diff
@@ -29705,7 +29708,20 @@ fn coerceExtra(
                 // T to E!T
                 return sema.wrapErrorUnionPayload(block, dest_ty, inst, inst_src) catch |err| switch (err) {
-                    error.NotCoercible => break :eu,
+                    error.NotCoercible => {
+                        const payload_type = dest_ty.errorUnionPayload(zcu);
+                        const intermediate = sema.coerceExtra(block, payload_type, inst, inst_src, .{ .report_err = false }) catch |intermed_err| switch (intermed_err) {
+                            error.NotCoercible => {
+                                if (in_memory_result == .no_match) {
+                                    // Try to give more useful notes
+                                    in_memory_result = try sema.coerceInMemoryAllowed(block, payload_type, inst_ty, false, target, dest_ty_src, inst_src, maybe_inst_val);
+                                }
+                                break :eu;
+                            },
+                            else => |e| return e,
+                        };
+                        return try sema.wrapErrorUnionPayload(block, dest_ty, intermediate, inst_src);
+                    },
```
No need for the second coercion here since `wrapErrorUnionPayload` is doing the exact same thing.
```suggestion
                    error.NotCoercible => {
                        if (in_memory_result == .no_match) {
                            const payload_type = dest_ty.errorUnionPayload(zcu);
                            // Try to give more useful notes
                            in_memory_result = try sema.coerceInMemoryAllowed(block, payload_type, inst_ty, false, target, dest_ty_src, inst_src, maybe_inst_val);
                        }
                        break :eu;
                    },
```

WillLillis replied:
Makes sense, thanks!
